File: container_planner/oog.py
from __future__ import annotations

from decimal import Decimal
from typing import Iterable

from container_planner.models import OogResult, Orientation, Piece, ContainerSpec
from container_planner.rounding import ceil_cm, ceil_m3


def _orientations(piece: Piece) -> list[Orientation]:
    dims = [piece.L_cm, piece.W_cm, piece.H_cm]
    keys = [
        (0, 1, 2, "LWH"),
        (0, 2, 1, "LHW"),
        (1, 0, 2, "WLH"),
        (1, 2, 0, "WHL"),
        (2, 0, 1, "HLW"),
        (2, 1, 0, "HWL"),
    ]
    seen = set()
    result = []
    for a, b, c, key in keys:
        oriented = (dims[a], dims[b], dims[c])
        if oriented in seen:
            continue
        seen.add(oriented)
        result.append(Orientation(L_cm=oriented[0], W_cm=oriented[1], H_cm=oriented[2], rotation_key=key))
    return result


def choose_orientation(piece: Piece) -> list[Orientation]:
    if piece.rotate_allowed:
        return _orientations(piece)
    return [Orientation(L_cm=piece.L_cm, W_cm=piece.W_cm, H_cm=piece.H_cm, rotation_key="LWH")]


def evaluate_oog(piece: Piece, ref: ContainerSpec) -> OogResult:
    if ref.inner_L_cm is None or ref.inner_W_cm is None or ref.inner_H_cm is None:
        raise ValueError("OOG判定にはSTANDARDコンテナ内寸が必要です")
    best = None
    for orientation in choose_orientation(piece):
        over_L = max(Decimal("0"), orientation.L_cm - ref.inner_L_cm)
        over_W = max(Decimal("0"), orientation.W_cm - ref.inner_W_cm)
        over_H = max(Decimal("0"), orientation.H_cm - ref.inner_H_cm)
        score = over_L + over_W + over_H
        if best is None or score < best[0]:
            best = (score, orientation, over_L, over_W, over_H)
    _, orientation, over_L, over_W, over_H = best
    over_L = ceil_cm(over_L)
    over_W = ceil_cm(over_W)
    over_H = ceil_cm(over_H)
    oog_flag = any(over > 0 for over in [over_L, over_W, over_H])
    if over_W > 0 or over_L > 0:
        suggestion = "FR"
    elif over_H > 0 and over_L == 0 and over_W == 0:
        suggestion = "OT"
    else:
        suggestion = ""
    protrude_L = ceil_m3((over_L * orientation.W_cm * orientation.H_cm) / Decimal("1000000"))
    protrude_W = ceil_m3((orientation.L_cm * over_W * orientation.H_cm) / Decimal("1000000"))
    protrude_H = ceil_m3((orientation.L_cm * orientation.W_cm * over_H) / Decimal("1000000"))

    door_check_applied = ref.door_W_cm is not None and ref.door_H_cm is not None
    door_passable = True
    door_over_w = Decimal("0")
    door_over_h = Decimal("0")
    door_reason = ""
    if door_check_applied:
        best_door = None
        for candidate in choose_orientation(piece):
            over_door_w = max(Decimal("0"), candidate.W_cm - ref.door_W_cm)
            over_door_h = max(Decimal("0"), candidate.H_cm - ref.door_H_cm)
            score = over_door_w + over_door_h
            if best_door is None or score < best_door[0]:
                best_door = (score, over_door_w, over_door_h)

        _, door_over_w, door_over_h = best_door
        door_over_w = ceil_cm(door_over_w)
        door_over_h = ceil_cm(door_over_h)
        door_passable = door_over_w == 0 and door_over_h == 0
        if not door_passable:
            reasons = []
            if door_over_w > 0:
                reasons.append(f"入口幅超過 {door_over_w}cm")
            if door_over_h > 0:
                reasons.append(f"入口高さ超過 {door_over_h}cm")
            door_reason = " / ".join(reasons)

    return OogResult(
        oog_flag=oog_flag,
        oog_ref_type=ref.type,
        over_L_cm=over_L,
        over_W_cm=over_W,
        over_H_cm=over_H,
        suggestion=suggestion,
        protrude_L_m3=protrude_L,
        protrude_W_m3=protrude_W,
        protrude_H_m3=protrude_H,
        chosen_orientation=orientation,
        door_passable=door_passable,
        door_check_applied=door_check_applied,
        door_over_W_cm=door_over_w,
        door_over_H_cm=door_over_h,
        door_reason=door_reason,
    )


def summarize_oog_overages(oog_results: list[tuple[Piece, OogResult]]) -> dict[str, Decimal]:
    ow_each = Decimal("0")
    oh = Decimal("0")
    for _, oog in oog_results:
        ow_each += oog.over_W_cm
        oh += oog.over_H_cm
    return {"OW_each": ow_each, "OH": oh}

File: web/src/domain/oog.ts
import type { ContainerSpec, OogResult, Orientation, Piece } from "./types";
import { ceilCm } from "./rounding";

export const orientationsFor = (piece: Piece): Orientation[] => {
  if (!piece.rotate_allowed) {
    return [{ L_cm: piece.L_cm, W_cm: piece.W_cm, H_cm: piece.H_cm, rotation_key: "LWH" }];
  }
  const dims = [piece.L_cm, piece.W_cm, piece.H_cm];
  const permutations: [number, number, number, string][] = [
    [0, 1, 2, "LWH"],
    [0, 2, 1, "LHW"],
    [1, 0, 2, "WLH"],
    [1, 2, 0, "WHL"],
    [2, 0, 1, "HLW"],
    [2, 1, 0, "HWL"],
  ];
  const seen = new Set<string>();
  return permutations.flatMap(([a, b, c, key]) => {
    const values = [dims[a], dims[b], dims[c]];
    const signature = values.join("/");
    if (seen.has(signature)) return [];
    seen.add(signature);
    return [{ L_cm: values[0], W_cm: values[1], H_cm: values[2], rotation_key: key }];
  });
};

export const evaluateOog = (piece: Piece, ref: ContainerSpec): OogResult => {
  const ranked = orientationsFor(piece)
    .map((orientation) => {
      const overL = Math.max(0, orientation.L_cm - ref.inner_L_cm);
      const overW = Math.max(0, orientation.W_cm - ref.inner_W_cm);
      const overH = Math.max(0, orientation.H_cm - ref.inner_H_cm);
      return { orientation, overL, overW, overH, score: overL + overW + overH };
    })
    .sort((a, b) => a.score - b.score);
  const best = ranked[0];
  const overL = ceilCm(best.overL);
  const overW = ceilCm(best.overW);
  const overH = ceilCm(best.overH);
  const doorCheck = ref.door_W_cm != null && ref.door_H_cm != null;
  let doorOverW = 0;
  let doorOverH = 0;
  if (doorCheck) {
    const doorBest = orientationsFor(piece)
      .map((orientation) => ({
        overW: Math.max(0, orientation.W_cm - (ref.door_W_cm ?? 0)),
        overH: Math.max(0, orientation.H_cm - (ref.door_H_cm ?? 0)),
      }))
      .sort((a, b) => a.overW + a.overH - (b.overW + b.overH))[0];
    doorOverW = ceilCm(doorBest.overW);
    doorOverH = ceilCm(doorBest.overH);
  }
  const doorReasons = [];
  if (doorOverW > 0) doorReasons.push(`入口幅超過 ${doorOverW}cm`);
  if (doorOverH > 0) doorReasons.push(`入口高さ超過 ${doorOverH}cm`);
  return {
    oog_flag: overL > 0 || overW > 0 || overH > 0,
    oog_ref_type: ref.type,
    over_L_cm: overL,
    over_W_cm: overW,
    over_H_cm: overH,
    suggestion: overL > 0 || overW > 0 ? "FR" : overH > 0 ? "OT" : "",
    chosen_orientation: best.orientation,
    door_passable: doorOverW === 0 && doorOverH === 0,
    door_check_applied: doorCheck,
    door_over_W_cm: doorOverW,
    door_over_H_cm: doorOverH,
    door_reason: doorReasons.join(" / "),
  };
};

const RF_KEYWORDS = ["reefer", "refrigerated", "frozen", "cold", "冷凍", "冷蔵", "要冷", "rf"];

export const requiresReefer = (piece: Piece): boolean => {
  const target = `${piece.desc} ${piece.package_text}`.toLowerCase();
  return RF_KEYWORDS.some((keyword) => target.includes(keyword));
};

export const isBreakbulkRequired = (piece: Piece, fortyFr?: ContainerSpec): boolean => {
  const deckLength = fortyFr?.deck_L_cm ?? 1160;
  const payload = fortyFr?.max_payload_kg ?? 34_000;
  const footprintFits = piece.L_cm <= deckLength || piece.W_cm <= deckLength;
  return !footprintFits || piece.weight_kg > payload;
};

const FR_MIN_FOOTPRINT_M2 = 2;
const FR_MIN_VOLUME_M3 = 2;
const FR_MIN_WEIGHT_KG = 2_000;

export const isFrCargoSubstantial = (piece: Piece): boolean => {
  const largestFootprintM2 = Math.max(
    piece.L_cm * piece.W_cm,
    piece.L_cm * piece.H_cm,
    piece.W_cm * piece.H_cm,
  ) / 10_000;
  return largestFootprintM2 >= FR_MIN_FOOTPRINT_M2 || piece.m3 >= FR_MIN_VOLUME_M3 || piece.weight_kg >= FR_MIN_WEIGHT_KG;
};
